text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,482 @@
+and detailed in [OTEP 0205](0205-messaging-semantic-conventions-context-propagation).
+
+* [Motivation](#motivation)
+* [Terminology](#terminology)
+* [Stages of producing and consuming messages](#stages-of-producing-and-consuming-messages)
+* [Trace structure](#trace-structure)
+* [Proposed changes and additions to the messaging semantic conventions](proposed-changes-and-additions-to-the-messaging-semantic-conventions)
+  - [Operation name](operation-name)
+  - [Span kind](span-kind)
+  - [Span relationships](span-relationships)
+* [Examples](#examples)
+* [Future possibilities](#future-possibilities)
+  - [Intermediary instrumentation](#intermediary-instrumentation)
+
+## Motivation
+
+Tracking the path of an individual message through a distributed system poses
+several challenges. Messaging systems allow for asynchronous workflows, which
+means that the stages of producing and consuming a message can be separated by
+a considerable time gap (this can be minutes, hours, or days). Furthermore, one
+cannot rely on consistent instrumentation across all parts of the system that
+touch a message. Correlating producer and consumer stages are expected even when
+the intermediary forwarding the message between them is not instrumented.
+Finally, batching of messages can happen in many different parts of a message
+processing workflow, be it batch publishing, batch receiving, batch processing,
+or batch settling.
+
+Despite all those challenges, requirements for instrumentation of messaging
+scenarios are high. Besides correlating spans that model the different
+processing stages of a message, it should also be possible to determine the
+end-to-end latency of processing a message. If intermediaries are not
+instrumented, this shouldn't impact the correlation of producer and consumer
+stages. If, on the other hand, intermediaries are instrumented, spans from
+intermediary instrumentation should seamlessly integrate with producer and
+consumer instrumentation.
+
+This OTEP aims at proposing consistent guidelines for creating spans that model
+the stages of the messaging flow, and correlating those in a way so that the
+requirements sketched above can be met in a consistent way across messaging
+scenarios and different messaging systems.
+
+## Terminology
+
+For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).
+
+## Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  CR[Create] --> PU[Publish]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  end
+  subgraph CONSUMER
+  direction TB
+  RE[Receive] --> PR[Process]
+  PR --> SE[Settle]
+  end
+  PU --> INTERMEDIARY
+  INTERMEDIARY --> RE
+  SE -..-> INTERMEDIARY
```

Comment:
Something that I raised in https://github.com/open-telemetry/oteps/pull/192 as well. We should have settlement on the producer side as well imho.

Reply:
Thanks for pointing this out. Could you give some use cases for producer settlement, or links to messaging systems that support it?

Reply:
AFAIK mostly "all" messaging systems have such a support? It's about getting an ACK that the message was received by the consumer (if peer-to-peer, but let's take this out of scope) or intermediary (broker) in order to allow different QoS (fire and forget/at most once, at least once, "exactly once").
For example, with AMQP 1.0 the sender sends the `transfer` frame, it can set `settled=true` flag already (so fire and forget) or `settled=false` and waiting for hearing back from the broker by receiving a `disposition` frame which settles the delivery (on the broker itself). So for example a messaging system like Apache Artemis has this.

Even the JMS specification has such a thing at API level (both with sync and asyn sends). See [async send](https://jakarta.ee/specifications/messaging/3.0/jakarta-messaging-spec-3.0.html#asynchronous-send) and more specifically this sentence.

> "In some Jakarta Messaging providers, a normal synchronous send involves sending the message to a remote Jakarta Messaging server and then waiting for an acknowledgement to be received before returning. It is expected that such a provider would implement an asynchronous send by sending the message to the remote Jakarta Messaging server and then returning without waiting for an acknowledgement. When the acknowledgement is received, the Jakarta Messaging provider would notify the application by invoking the onCompletion method on the application-specified CompletionListener object. If for some reason the acknowledgement is not received the Jakarta Messaging provider would notify the application by invoking the CompletionListener’s onException method."

Of course Kafka, the producer can set different levels of acks (not getting ack, getting ack from leader, getting ack when message is received from leader and replicated to followers). See official doc [here](https://kafka.apache.org/documentation/#producerconfigs_acks).

Reply:
Thanks for giving the links. We discussed this in today's workgroup meeting. To summarize, "producer settlement", according to our understanding, involves receiving an ACK on the producer side from the intermediary, confirming that the intermediary successfully received the message.

We identified two main scenarios, for modelling both we intend to look at it more from the user perspective than from the underlying messaging system perspective (like AMQP details, which will be covered by transport layer instrumentation).
1. For synchronous settlement, the user will initiate a publish operation, this operation will block until an ACK for each message is received. This should be covered by a single "Publish" span.
2. For asynchronous settlement, the user will initiate a publish operation, which returns before awaiting an ACK from the intermediary. The instrumentation in this case should reflect the way that status is returned to the user. If a callback is invoked, this could involve a separate span for this callback which is parented/linked to the initial "Publish" span. However, we think the possibilities here are very varied and bound to implementations of messaging SDKS, so we don't want to come up with generic conventions for those cases at this point.

Reply:
> We discussed this in today's workgroup meeting

And I had conflicts as usual on my agenda. Sorry but there is a weekly meeting I have that time which is never cancelled :-(

Regarding the point 1 and 2 ...
How are you going to differentiate scenario 1 where sync message was sent, you have a "Publish" span only because an ACK was returned compared to scenario 2 where you have async message set, you have a "Publish" span  but the user asked for "fire and forget" so no ACK was got by the SDK. Won't they overlap? 
Instead, when a callback is called are you referring to create a "Settle" span linked to the "Publish"?

Reply:
This was discussed in today's messaging workgroup. The consensus in the group was, that also asynchronous settlement operations on the producer side should be modelled with a single span whenever possible, this span should end when the ACK was received.

In cases where this is not possible, a using a separate "settle" span makes sense. However, as this seems to be an edge case, we decided not to add additional complexity by specifying this but leave more flexibility to instrumentors.